storm-client/src/jvm/org/apache/storm/daemon/Task.java

```diff
@@ -122,28 +130,33 @@ public Task(Executor executor, Integer taskId) throws IOException {
         return new ArrayList<>(0);
     }
 
+
     public List<Integer> getOutgoingTasks(String stream, List<Object> values) {
         if (debug) {
             LOG.info("Emitting Tuple: taskId={} componentId={} stream={} values={}", taskId, componentId, stream, values);
         }
 
-        List<Integer> outTasks = new ArrayList<>();
-        if (!streamComponentToGrouper.containsKey(stream)) {
-            throw new IllegalArgumentException("Unknown stream ID: " + stream);
-        }
-        if (null != streamComponentToGrouper.get(stream)) {
-            // null value for __system
-            for (LoadAwareCustomStreamGrouping grouper : streamComponentToGrouper.get(stream).values()) {
+        ArrayList<Integer> outTasks = new ArrayList<>();
+
+        // TODO: PERF: expensive hashtable lookup in critical path
```

Comment:
Pretty much all these todos are intended to get addressed in this PR itself. Just wanted to call them out during this review. There are maybe 2 perf todo observations that I am hoping to leave in alongside the offending code ..  as it took a lot of time to identify them (but not easy to fix here). So valuable to have them in IMO.